pkg/metrics/garbage_collector.go

```diff
@@ -0,0 +1,190 @@
+package metrics
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	nm "github.com/rancher/norman/metrics"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+var (
+	GCTargetMetricsByNameForClientKey = []interface{}{
+		nm.TotalAddWS, nm.TotalRemoveWS, nm.TotalAddConnectionsForWS, nm.TotalRemoveConnectionsForWS,
+		nm.TotalTransmitBytesOnWS, nm.TotalTransmitErrorBytesOnWS, nm.TotalReceiveBytesOnWS,
+	}
+
+	GCTargetMetricsByIPForPeer = []interface{}{
+		nm.TotalAddPeerAttempt, nm.TotalPeerConnected, nm.TotalPeerDisConnected,
+	}
+)
+
+func MetricGarbageCollection(context *config.ScaledContext) {
+	logrus.Debugf("MetricGarbageCollector Start")
+
+	clusterLister := context.Management.Clusters("").Controller().Lister()
+	nodeLister := context.Management.Nodes("").Controller().Lister()
+	endpointLister := context.Core.Endpoints(settings.Namespace.Get()).Controller().Lister()
+	isClusterMode := settings.Namespace.Get() != "" && settings.PeerServices.Get() != ""
+
+	// Existing resources
+	observedResourceNames := map[string]bool{}
+	// Exisiting Metrics Labels
+	observedLabelsMap := map[string]map[interface{}][]map[string]string{}
+
+	// Get Clusters
+	clusters, err := clusterLister.List("", labels.Everything())
+	if err != nil {
+		logrus.Errorf("MetricGarbageCollector failed to list clusters: %s", err)
+		return
+	}
+	for _, cluster := range clusters {
+		if _, ok := observedResourceNames[cluster.Name]; !ok {
+			observedResourceNames[cluster.Name] = true
+		}
+	}
+	// Get Nodes
+	nodes, err := nodeLister.List("", labels.Everything())
+	if err != nil {
+		logrus.Errorf("MetricGarbageCollector failed to list nodes: %s", err)
+	}
+	for _, node := range nodes {
+		if _, ok := observedResourceNames[node.Namespace+":"+node.Name]; !ok {
+			observedResourceNames[node.Namespace+":"+node.Name] = true
+		}
+	}
+	// Get Endpoints
+	endpoints, err := endpointLister.List(settings.Namespace.Get(), labels.Everything())
+	countOfRancherServer := 0
+	if isClusterMode {
```

Comment:
`isClusterMode` variable judge if rancher-server is running cluster mode or not like https://github.com/rancher/rancher/blob/a0b4e1b1f363d906a8daebabf65f0e9a5fc96482/pkg/tunnelserver/peermanager.go#L41.

Only when rancher-server is running as a cluster-mode, we need to collect all of rancher-server's IP address from endpoints resource as a observed resources since session server will keep websocket session for all of rancher-servers as a peering target.

Later this observed resources will be used when we judge if this metrics is still used or not. for example, if rancher-server's ip address has been changed, we don't want to keep metrics for old peer.

Let me consider how we can simplify this logic